paddle/operators/margin_rank_loss_op.cc

```diff
@@ -0,0 +1,114 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/operators/margin_rank_loss_op.h"
+
+namespace paddle {
+namespace operators {
+
+class MarginRankLossOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    // input check
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Label"),
+                            "Input(Label) shouldn't be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X1"), "Input(X1) shouldn't be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X2"), "Input(X2) shouldn't be null");
+    auto label_dims = ctx.Input<framework::Tensor>("Label")->dims();
+    auto x1_dims = ctx.Input<framework::Tensor>("X1")->dims();
+    auto x2_dims = ctx.Input<framework::Tensor>("X2")->dims();
+    PADDLE_ENFORCE((label_dims == x1_dims) && (x1_dims == x2_dims) &&
+                       (label_dims.size() == 2) && (label_dims[1] == 1),
+                   "All inputs must be vector with the same size");
+    ctx.Output<framework::LoDTensor>("Activated")->Resize(label_dims);
+    ctx.Output<framework::LoDTensor>("Out")->Resize(label_dims);
```

Comment:
Should here add the following codes?  I am not sure, because for this operator the input X1, X2, and the output are always non-sequence. In this case, are the codes below still necessary? @qingqing01

```cpp
ctx.ShareLoD("Out", /*->*/ "X1");
```

Reply:
Maybe not necessary here